### DISCUSSION TITLE
How should Dropout be used?

### DISCUSSION THREAD

**Discussion by blackblitz** [google/flax#5553] [Q&A]
[Flax Basics](https://flax.readthedocs.io/en/latest/nnx_basics.html#nested-modules) says "we need to pass the `__call__` method the RNG state that we want the `Dropout` layer to use." However, the API docs for `Dropout` shows an example usage without passing the RNG state to the `__call__` method:

```python
class MLP(nnx.Module):
  def __init__(self, rngs):
    self.linear = nnx.Linear(in_features=3, out_features=4, rngs=rngs)
    self.dropout = nnx.Dropout(0.5, rngs=rngs)
  def __call__(self, x):
    x = self.linear(x)
    x = self.dropout(x)
    return x
```

What is the recommended way to use `Dropout`? It should be clearly written in the docs.

**Comment by MLuc24** [comment]
Both are correct — they're describing two different setups, not disagreeing.

`Dropout` takes `rngs` in the constructor *or* at call time, and resolves them in that order at call:

```python
rngs = first_from(
  rngs,
  self.rngs,
  error_msg="""`deterministic` is False, but no `rngs` argument was provided to Dropout
      as either a __call__ argument or class attribute.""",
)
```

so passing `rngs` to `__init__` (as the API docs example does) means there's nothing left to pass at `__call__`. The Basics page is describing the other case — a `Dropout` constructed without `rngs`, where the call has to supply them. If neither does, you get that error rather than silent wrong behaviour.

The recommended form is the one in the API docs: give it `rngs` at construction, then control it with `model.train()` / `model.eval()`, which flip `deterministic` for you. Passing `rngs` at call time is the escape hatch for when you want a specific key for a specific call — the docstring notes it "will take precedence over the rngs passed into the constructor".

Worth filing the doc wording separately though. The Basics sentence reads as a general requirement rather than a description of one setup, which is exactly how you read it.
